### PR TITLE
Wire new dqlite implementation

### LINKED ISSUE #4548: LXD cluster 'lxc list' command extremely laggy
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
```

config:
  core.https_address: 10.1.2.41:8443
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 10.1.2.41:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFTTCCAzWgAwIBAgIRAIYi9SOBaeGArc8ha+7ZyvIwDQYJKoZIhvcNAQELBQAw
    ODEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEYMBYGA1UEAwwPcm9vdEBs
    eGQtZGV2LTAxMB4XDTE4MDUwODIwNTU1N1oXDTI4MDUwNTIwNTU1N1owODEcMBoG
    A1UEChMTbGludXhjb250YWluZXJzLm9yZzEYMBYGA1UEAwwPcm9vdEBseGQtZGV2
    LTAxMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA0JxeMil0dOsHwO/x
    xIHJHXfC7VlNRP0WXNLb9glwUr3qVEExW+NG5XOdzts5uQm0wV1nhDrZoLr3AUwI
    dmrvREWbn7PQEr0Fw7ZHYPG1r+06b5m0fweEWuV6punvzZSCeiIV5M4w6S+93Xna
    C4fryXuIRIpMrXhhVzSclIGWiB4DA95KLN8oR8FfhEstZ6Jz+sYMMrOxDjbhEYLw
    OSV4MYgFhXlA+KGgeyCOrseiz9hycBeM8zniIxDzFIFTSGWd9B3omb4eK2li6zhb
    FNnNqsCdIOY7HsMJyD2bLJCKsuxXYkpN/XVJCNVOEoU13S1arTnIKPVzXvnFBpZx
    BREIbBQRI5fROYqlrbcMvOq/jEoWD9JQobK0OhfvdrTaJpWENF2ttf4QwANPx/zC
    DtSfIUbJ5I7iP0P+IuaShgwM2LG0aqyEwEpP3lPqV/LAPFiLj77QuxudiFVU0yZO
    iaA2UUFOskTNQNiMdZqD87j2Pp74dDrKSvmxTSwlXwJntUxAKoSUzaMBGhiRsuGp
    uc3atWRG+Ne7BMpepuyvEb2+kmFjexiyqH1pAXaAhNPJHithCiTFaOJQjbHx+n1n
    AxWqGnJQoDTyRm/JA1kyBo66NoDGW5X/RjVYIkD3Dc1svkYP6eWOB0Mi1mYj0qaq
    LG8jZ/e71ak9GSgdHDzXXvrLP0kCAwEAAaNSMFAwDgYDVR0PAQH/BAQDAgWgMBMG
    A1UdJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwGwYDVR0RBBQwEoIKbHhk
    LWRldi0wMYcECgECKTANBgkqhkiG9w0BAQsFAAOCAgEAPTKfYKvM4tNuoItiIZ7C
    lKVN+73FriGg+xkl4APuT/FqciP4rXQ4eI+XW81BBc4jx1kK5ngmZbOO8tXROBZk
    lHp/6Vz1gxYLeLSbZA5Pi6dnJE5zJnPGrB3vUQgVxdlIF1zI2jtQ9h1h0wUH97nB
    uoK2f0Wd4S+z10bac6P4wGqrzwtFsfnT+077EIdpgtXeqQOxId1GaIfWEKzUhCRO
    Aj6aYw0bU5bigD0OhROI246nKk363X+4kZyMT7FNpUZMpESrw/ZjCBaCOkZkqOEq
    huCVn4yWhZ1sYIMlg9aitNeKaU70gOZDHajI+Lo8m2eemmiQoWGUqN8vu8Tk0Kzo
    s4bFhjPv/NFd9h2LvuulhBqmclPszgyhk9CrSPJL0K0GgcVUMG3IiuxoHBqVQn9M
    nPOVPlf2cYqb4OFC+kdDT3JqG29K+4am8gvV6rAFl+dCsRd6k45keWAB1MjQfl7N
    8whAL6TEXmiT1sb3pWvgd7E53SeIAXEPP/furs0excMJvu/yfZOGhaD/NTZzoC8g
    5WvNNxuuvuWISMJL0uHrn6NcgZ1rZFeOdMwOYewElyEkVLbFMpTacrGj3vYlIEfK
    MXSIiGOGMundoCvlDRRcLCqkYGbiu2Nlab23wpMWBxpFASsaZGgEXRRIn4ujie3t
    OboWBqb/G5F+82vh6wgV2t0=
    -----END CERTIFICATE-----
  certificate_fingerprint: fcc5ff4aea9e3513c7e2791e865bdae6f81b321c6c007a2fd7498d1d727060d0
  driver: lxc
  driver_version: 3.0.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.15.0-20-generic
  server: lxd
  server_pid: 2245
  server_version: 3.0.0
  storage: zfs
  storage_version: 0.7.5-1ubuntu13
  server_clustered: true
  server_name: lxd-dev-01
```
# Issue description

When running ` lxc list ` the time to return information on 11 containers takes anywhere from 15 - 60 seconds. I understand that there will be some extra laggyness on clusters because of the quorum database, but I feel like something is going wrong or timing out on the back-end.
[lxc-list-debug.txt](https://github.com/lxc/lxd/files/1989229/lxc-list-debug.txt)

# Steps to reproduce

 1. Set up 3 node cluster on same subnet, no maas server or fan networking involved.
 2. run `lxc list` on any of the nodes
```
ubuntu@lxd-dev-01:~$ time lxc list

real	0m28.707s
user	0m0.088s
sys	0m0.045s
```

The above command was run right after a previous command that seemed to take even longer. It is consistently slow.

# Information to attach
I managed to capture the debug output of one example taking over 1 minute to list out information about 11 containers.

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [x] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

### PR DESCRIPTION
**Pull request by freeekanayaka**
This branch wires the new Go bindings for dqlite into the lxd code base. The commits are mostly organized by sub-package, I believe it should not be too bad to backport.

This fixes #4548 as far as the low-level dqlite code is concerned. We'll probably want to keep improving the situation over time by making db-related code of LXD itself more efficient (use prepared statements, use views, minimize round trips, etc) and looking at other spots where we can speed up lxc list.

### CODE REVIEW

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Doesn't run on 32bit arches.

**Comment by stgraber** [comment]
Error messages need to be updated to match our pattern of starting with a capital and not ending with a period

**Comment by stgraber** [comment]
As for the DB being flushed to disk thing, can we add a `.sync` command or something to `lxd sql` and use that instead of having to either skip the empty table checks or have to bounce the daemon to have it sync the database?

**Comment by stgraber** [comment]
Also, should I kill the go-grpc-sql and go-sqlite3 repos in the Canonical namespace?

**Comment by stgraber** [comment]
Those two runners show the same error:
 - https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/4625/arch=amd64,backend=lvm,compiler=golang-1.10,restrict=lxd-priv/console
 - https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/4625/arch=amd64,backend=zfs,compiler=golang-1.10,restrict=lxd-priv/console

```
==> Spawned LXD (PID is 243993)
==> Confirming lxd is responsive
INFO[07-30|23:22:43] LXD 3.3 is starting in normal mode       path=/lxc-ci/build/tmp.iTEVMYRIt7/go/src/github.com/lxc/lxd/test/tmp.Git/Ktw
INFO[07-30|23:22:43] Kernel uid/gid map: 
INFO[07-30|23:22:43]  - u 0 0 4294967295 
INFO[07-30|23:22:43]  - g 0 0 4294967295 
INFO[07-30|23:22:43] Configured LXD uid/gid map: 
INFO[07-30|23:22:43]  - u 0 100000 1000000000 
INFO[07-30|23:22:43]  - g 0 100000 1000000000 
WARN[07-30|23:22:43] CGroup memory swap accounting is disabled, swap limits will be ignored. 
INFO[07-30|23:22:43] Initializing local database 
INFO[07-30|23:22:43] Updating the LXD database schema. Backup made as "local.db.bak" 
Error: LXD still not running after 300s timeout (Get http://unix.socket/1.0: dial unix /lxc-ci/build/tmp.iTEVMYRIt7/go/src/github.com/lxc/lxd/test/tmp.Git/Ktw/unix.socket: connect: no such file or directory)
```

So looks like a hang on initial startup, this needs to be tracked down as it hit two different machines using two different storage backends on the first test run, so chances are we'll see a lot more of that.

**Comment by freeekanayaka** [comment]
I capitalized all error messages introduced or modified by this branch.

**Comment by freeekanayaka** [comment]
I added a the "lxd sql global .sync" command since it seems a good idea in general.

However, I think that the fact that the test for activateifneeded now actually bounces the daemon is a good thing because it ensures that the shutdown logic does sync to disk. If we were to manually run .sync in the test, we might miss a regression in the shutdown logic.

**Comment by freeekanayaka** [comment]
We can surely kill go-grpc-sql. As for go-sqlite3, we currently don't needed but if we want "go get" to work without passing the "-tags libsqlite3" flag, we probably have to keep it. I don't like it much, because CanonicalLtd/go-sqlite3 would be essentially a copy of mattn/go-sqlite3 but with the vendored sqlite3 code replaced with the one that contains our patch. Also, in order to make "get run" I'll need to vendor libuv as well (in go-dqlite). I'm not entirely sure we want to go down this path, but I don't see other alternatives if we want "go get" to work.

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
Regarding the two jobs that got stuck during startup, I believe that's fixed now.

**Comment by freeekanayaka** [comment]
I've re-enabled the empty table checks also for the clustering case. It seems to actually work properly. But let's see what CI has to say.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/cluster/gateway.go [hunk 1]
```diff
@@ -409,28 +454,38 @@ func (g *Gateway) init() error {
 	// should serve as database node, so create a dqlite driver to be
 	// exposed it over gRPC.
 	if raft != nil {
-		config := dqlite.DriverConfig{}
-		driver, err := dqlite.NewDriver(raft.Registry(), raft.Raft(), config)
+		listener, err := net.Listen("unix", "")
 		if err != nil {
-			return errors.Wrap(err, "failed to create dqlite driver")
+			return errors.Wrap(err, "failed to allocate loopback port")
```
Comment (brauner) [inline lxd/cluster/gateway.go]: s/failed/Failed/

File: lxd/cluster/gateway.go [hunk 2]
```diff
@@ -582,3 +667,41 @@ func dqliteLog(configuredLevel string) func(level, message string) {
...
+			panic(err)
+		}
+		go func() {
+			io.Copy(eagain.Writer{Writer: dst}, eagain.Reader{Reader: src})
+			src.Close()
+		}()
+		go func() {
+			io.Copy(eagain.Writer{Writer: src}, eagain.Reader{Reader: dst})
+			dst.Close()
+		}()
```
Comment (brauner) [inline lxd/cluster/gateway.go]: Maybe we want to log errors in here when `io.Copy()` exits?

**Comment by stgraber** [comment]
looking good but lets make sure

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by brauner** [comment]
Thanks everyone!

**Comment by brauner** [comment]
Thanks!

**Pull request merged**